### PR TITLE
hard max limit for slices

### PR DESCRIPTION
**Pull request by lozenko**
In some cases using ForLoop can led to OOM when we deal with many elements because we start to create lists with huge capacity. 

In this PR we introduce MAX_SLICES limit. Additional improvement is to relax the heap usage by not copying elements to new `ArrayLists` rather reusing the one in `ForLoop` which has `Iterator<?>` it inside.

### CODE REVIEW

File: src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java [hunk 1]
```diff
@@ -80,20 +80,26 @@ public Object filter(Object var, JinjavaInterpreter interpreter, String... args)
...
     int i = 0;
     while (loop.hasNext()) {
-      Object next = loop.next();
       if (i % slices == 0) {
-        currentList = new ArrayList<>(slices);
-        result.add(currentList);
+        if (currentList != null) {
+          result.add(currentList);
+        }
+        currentList = new ArrayList<>();
```
Comment (boulter) [inline src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java:93]:
```suggestion
        currentList = new ArrayList<>(slices);
```
so we never have to resize the array.

**Review by boulter — COMMENTED** [review]
Hard to tell if this will solve the problem. May need to run a test case through a profiler and check the memory allocation.

File: src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java (hunk 1 above)
Comment (lozenko) [reply src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java:102]: as we discussed we keep it due to better heap results

**Comment by lozenko** [comment]
added hard max limit

File: src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java [hunk 2]
```diff
@@ -79,21 +83,32 @@ public Object filter(Object var, JinjavaInterpreter interpreter, String... args)
         0,
         args[0]
       );
+    } else if (slices > MAX_SLICES) {
+      ENGINE_LOG.warn(
```
Comment (boulter) [inline src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java:87]: This won't be viewable by developers. We'd need to add a warning to the context like https://github.com/HubSpot/jinjava/blob/0de0e078e14e210719c622a6032621bf5c5e49af/src/main/java/com/hubspot/jinjava/lib/filter/PlusTimeFilter.java#L64

Comment (lozenko) [reply src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java]: replaced by your suggestion

File: src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java [hunk 3]
```diff
@@ -79,21 +82,41 @@ public Object filter(Object var, JinjavaInterpreter interpreter, String... args)
         0,
         args[0]
       );
+    } else if (slices > MAX_SLICES) {
+      interpreter.addError(
+        new TemplateError(
+          TemplateError.ErrorType.WARNING,
+          TemplateError.ErrorReason.OVER_LIMIT,
+          TemplateError.ErrorItem.FILTER,
+          "The limit input value is too large, it's been reduced to " + MAX_SLICES,
```
Comment (boulter) [inline src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java:91]: Can you refer to the `slices` count as described in the docs on line 26?
Also please add the 1000 limit to those docs.

**Review by boulter — COMMENTED** [review]
One comment and then I think we're good to merge.

File: src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java (hunk 3 above)
Comment (lozenko) [reply src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java]: added

File: src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java [hunk 4]
```diff
@@ -79,21 +86,42 @@ public Object filter(Object var, JinjavaInterpreter interpreter, String... args)
         0,
         args[0]
       );
+    } else if (slices > MAX_SLICES) {
+      interpreter.addError(
+        new TemplateError(
+          TemplateError.ErrorType.WARNING,
+          TemplateError.ErrorReason.OVER_LIMIT,
+          TemplateError.ErrorItem.FILTER,
+          "The limit input value of 'slices' parameter is too large, it's been reduced to " +
```
Comment (boulter) [inline src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java:95]:
```suggestion
          "The value of the 'slices' parameter is greater than " + MAX_SLICES + ". It's been reduced to " +
```

File: src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java [hunk 5]
```diff
@@ -25,7 +26,11 @@
     @JinjavaParam(
       value = "slices",
       type = "number",
-      desc = "Specifies how many items will be sliced",
+      desc = "Specifies how many items will be sliced." +
+      "Please note, the max limit is " +
+      SliceFilter.MAX_SLICES +
+      ". " +
+      "All values above this limit will be reduced.",
```
Comment (boulter) [inline src/main/java/com/hubspot/jinjava/lib/filter/SliceFilter.java:33]:
```suggestion
      desc = "Specifies how many items will be sliced. Maximum value is " +
      SliceFilter.MAX_SLICES +
      ". ",
```

**Review by boulter — COMMENTED** [review]
Simplified the language a little here.

**Review by boulter — APPROVED** [review]

**Pull request merged**